core/src/main/java/com/linecorp/armeria/common/RequestContext.java

```diff
@@ -596,60 +588,61 @@ default Logger makeContextAware(Logger logger) {
     }
 
     /**
-     * Registers {@code callback} to be run when re-entering this {@link RequestContext}, usually when using
-     * the {@link #makeContextAware} family of methods. Any thread-local state associated with this context
-     * should be restored by this callback.
+     * Do not use this method. The callback is never invoked.
      *
-     * @param callback a {@link Consumer} whose argument is this context
-     */
-    void onEnter(Consumer<? super RequestContext> callback);
-
-    /**
-     * Registers {@code callback} to be run when re-entering this {@link RequestContext}, usually when using
-     * the {@link #makeContextAware} family of methods. Any thread-local state associated with this context
-     * should be restored by this callback.
+     * <p>This method has a limitation which is that the callback cannot be called when this
+     * {@link RequestContext} is pushed first. So we decided to remove this method and provide
+     * better API after 1.0. If you are using this, please find another way to achieve the same behavior.
+     * Most of the cases, you might be able to just use {@link RequestLogListener}.
+     * Also, you might follow this approach which is using your own thread-local.
+     * https://github.com/SourceForgery/tachikoma/pull/191
+     * If you're still struggle to replace this method, please come to our slack workspace and let us know.
+     * We might be a help to you.
+     *
+     * <p>Please check out https://github.com/line/armeria/pull/2375
      *
-     * @deprecated Use {@link #onEnter(Consumer)} instead.
+     * @deprecated Do not use this method.
      */
     @Deprecated
-    default void onEnter(Runnable callback) {
-        onEnter(ctx -> callback.run());
+    default void onEnter(Consumer<? super RequestContext> callback, String doNotUseThisMethod) {
+        /* no-op */
```

Comment:
How about just throwing `UnsupportedOperationException` with some message instead of adding a parameter?

Reply:
`UnsupportedOperationException` is a good choice.
Or how about making breaking changes in this case by removing these methods? 
Because 1) I guess @minwoox added `String doNotUseThisMethod` for making breaking changes on purpose to produce compile errors and 2) Users can easily detect that it does not work anymore.

Reply:
I agree that if making a behavior breaking change anyways it's better to just remove the API. We want to deprecate only where users will be able to continue to use their unmodified code.

Reply:
Just wanted to give more information from Javadoc, but I guess it's too much. Let me remove this. Thanks!